README.md
# Setup

0. (Optional) Install [CUDA](https://developer.nvidia.com/cuda-downloads) to enable inference with gpu (nvidia gpu only)

1. Conda environment
    1. Install [Miniconda](https://docs.conda.io/projects/miniconda/en/latest/index.html)
    2. Launch the miniconda terminal and setup an environment (in Windows should be a new app called "Anaconda Prompt (miniconda3)"):

        `conda create -n traffic_count python=3.8 -y`

        - "traffic_count" will be the name of the environment, feel free to change it, just remember to use the same name later in steps 2.1 and 3.1.
            - You can check if you already have a working environment with the following command:

                `conda info --envs`

        - "python=3.8" means that we will be using version 3.8 in our python environment. The project should work with versions 3.7 and 3.8, any higher than that might require some fixes on the code to keep it up to date.

2. Repositories setup

    1. Activate the environment made in step 1 using the miniconda terminal:
    
        `conda activate traffic_count`

    2. Install git:

        - Option 1: Install [Git](https://git-scm.com/downloads) for your operating system (restart your terminal and reactivate your environment after installation using this approach)

        - Option 2: Install git through conda:
        
            `conda install -c anaconda git`
    
    3. Clone repositories:

        1. Traffic-count-cctval:

            `git clone https://github.com/ijorquera/Traffic-Count-CCTVAL`
        
        2. torchreid:

            `git clone https://github.com/KaiyangZhou/deep-person-reid.git`

    4. (Jetson-nano) Install Traffic-count-cctval dependencies:

        1. Install requirements:

            `python3.8 -m pip install --no-cache-dir -r Traffic-Count-CCTVAL/requirements.txt`

        
        2. Install [pytorch](https://pytorch.org/get-started/locally/) with support for CUDA 10.2 (highest cuda version available for the Jetson nano)

            `python3.8 -m pip install torch==1.12.1+cu102 torchvision==0.13.1+cu102 torchaudio==0.12.1 --extra-index-url https://download.pytorch.org/whl/cu102`
            

    5. (Windows) Install Traffic-count-cctval dependencies:

        1. Install requirements:

            `pip install --no-cache-dir -r Traffic-Count-CCTVAL/requirements.txt`

        
        2. Install [pytorch](https://pytorch.org/get-started/locally/) (check on [pytorch](https://pytorch.org/get-started/locally/) for the right command for your system)

            - For latest version with cuda 12.1 support on windows:
            
                `pip install --no-cache-dir torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cu121`
            
            - For latest version with cpu only on windows:

                `pip install torch torchvision torchaudio`

    6. (Linux) Install Traffic-count-cctval dependencies:

        `pip install --no-cache-dir -r Traffic-Count-CCTVAL/requirements-linux.txt`

        - For linux, just requirements is enough since it will auto-detect wether there is a gpu with cuda support or not


    7. Install torchreid:

        `cd deep-person-reid`

        `python setup.py develop`

        `cd ..`
    
3. Traffic-count-cctval test:

    1. Activate the environment made in step 1 using the miniconda terminal:
    
        `conda activate traffic_count`
    
    2. Change directory to the repo with the project:

        `cd Traffic-Count-CCTVAL`
    
    3. Test with one of the sample videos (change `python` for `python3.8` on the Jetson nano):

        `python track.py --source heavy.mp4 --save-vid --show-vid --save-txt`

    4. Test with integrated camera:

        `python track.py --source 0 --save-vid --show-vid --save-txt`

4. Known errors:

    1. Error Placeholder


---

1. Media source
```
$ python track.py --source 0  # webcam
                           img.jpg  # image
                           vid.mp4  # video
                           path/  # whole directory
                           'https://youtu.be/link'  # YouTube
                           'rtsp://example.com/media.mp4'  # RTSP, RTMP, HTTP stream
```
2. Yolo Model
```
$ python track.py --source 0 --yolo_model yolov5n.pt 
                                          yolov5s.pt 
                                          yolov5m.pt       
                                          yolov5l.pt 
                                          yolov5x.pt 
                                          ...
```
3. Image size (pixels)
```
$ python track.py --source 0 --yolo_model yolov5n.pt --img 640  
                                                     --img 1280 
                       
```
4. How to save output
```
$ python track.py --source 0 --yolo_model yolov5n.pt --img 640  --save-vid
                                                                --show-vid
                                                                --save-txt
```


CHANGES.md
# Cambios realizados

- Se agregaron instrucciones de instalación/ejecución en el readme

- Limpieza del repositorio:
    - Se eliminó el cache de ejecuciones anteriores
    - Se agregó un .gitignore para evitar que se suban archivos muy pesados al repositorio (pesos del modelo, cache, entre otros)

- Cambios al archivo track.py:
    - Se agregaron líneas de código (líneas 215-219) para que el output se guarde en un archivo de texto (csv) siguiendo un formato más interpretable
        - El formato del output es el siguiente:
            `indice, número del frame, etiqueta, número del elemento según etiqueta, bbox_left, bbox_top, bbox_width, bbox_height`

            - número del frame: identificador para saber que frame de la fuente (video, cámara, imágen) se procesó.
            - etiqueta: categoría inferida por el programa (i.e. camión, bus, bicicleta, etc). De momento la etiqueta incluye el "número del elemento", la "categoría" o "clase", y la probabilidad inferida por el modelo.
            - número del elemento: identificador único que se le asigna a cada elemento detectado. El primer auto tendrá el número uno, el segundo el número dos, luego la primera bicicleta tendrá el número tres, y así, la idea es evitar contar dos veces el mismo elemento detectado, si se vuelve a encontrar un elemento que el modelo ya vió le asignará el mismo número que le asignó inicialmente y no lo considerará para el conteo. El mayor "número del elemento" dentro del archivo de salida indicará la cantidad de vehículos/elementos únicos contados.
            - bbox_left, bbox_top: coordenadas en pixeles para el punto superior izquierdo del cuadro que encierra la clase detectada.
            - bbox_width, bbox_height: ancho y alto en pixeles del cuadro que encierra la clase detectada.

---

# Pendiente:
- Revisar contador, eliminar clases
- Formato:
    - Ver el tema de las lineas horizontales de conteo
    - Dejar las etiquetas en español
    - Definir formato de entrega
    - Posibilidad de usar camara externa (usb u otros)

- Funcionalidad:
    - Revisar con videos que apunten a la calle en distintos ángulos

- Deploy:
    - Configurar repositorio para ser lanzado en múltiples fpga/sistemas
        - Beaglebone ai 64
        - Xilinix kria kv 260
        - Jetson nano

- Trabajo futuro:
    - Considerar alternativas para la detección de vehículos (Yolov8, segment-everything (facebook), otros)


main.py
import cv2
import os
import argparse
from network_model import model
from aux_functions import *

# Suppress TF warnings
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"

mouse_pts = []


def get_mouse_points(event, x, y, flags, param):
    # Used to mark 4 points on the frame zero of the video that will be warped
    # Used to mark 2 points on the frame zero of the video that are 6 feet away
    global mouseX, mouseY, mouse_pts
    if event == cv2.EVENT_LBUTTONDOWN:
        mouseX, mouseY = x, y
        cv2.circle(image, (x, y), 10, (0, 255, 255), 10)
        if "mouse_pts" not in globals():
            mouse_pts = []
        mouse_pts.append((x, y))
        print("Point detected")
        print(mouse_pts)


# Command-line input setup
parser = argparse.ArgumentParser(description="SocialDistancing")
parser.add_argument(
    "--videopath", type=str, default="vid_short.mp4", help="Path to the video file"
)
args = parser.parse_args()

input_video = args.videopath

# Define a DNN model
DNN = model()
# Get video handle
cap = cv2.VideoCapture(input_video)
height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
fps = int(cap.get(cv2.CAP_PROP_FPS))

scale_w = 1.2 / 2
scale_h = 4 / 2

SOLID_BACK_COLOR = (41, 41, 41)
# Setuo video writer
fourcc = cv2.VideoWriter_fourcc(*"XVID")
output_movie = cv2.VideoWriter("deteccion1.avi", fourcc, fps, (width, height))
bird_movie = cv2.VideoWriter(
    "visi1.avi", fourcc, fps, (int(width * scale_w), int(height * scale_h))
)
# Initialize necessary variables
frame_num = 0
total_pedestrians_detected = 0
total_six_feet_violations = 0
total_pairs = 0
abs_six_feet_violations = 0
pedestrian_per_sec = 0
sh_index = 1
sc_index = 1
# Read logo and resize
logo = cv2.imread('logo.png')
size1 = 100
size2 =150
logo = cv2.resize(logo, (size2, size1))
img2gray = cv2.cvtColor(logo, cv2.COLOR_BGR2GRAY)
ret, mask = cv2.threshold(img2gray, 1, 255, cv2.THRESH_BINARY)
cv2.namedWindow("image")
cv2.setMouseCallback("image", get_mouse_points)
num_mouse_points = 0
first_frame_display = True

# Process each frame, until end of video
while cap.isOpened():
    frame_num += 1
    ret, frame = cap.read()
            # Region of Image (ROI), where we want to insert logo
    roi = frame[-size1-10:-10, -size2-10:-10]

        # Set an index of where the mask is
    roi[np.where(mask)] = 0
    roi += logo


    if not ret:
        print("Final del procesamiento, TD 2021")
        break

    frame_h = frame.shape[0]
    frame_w = frame.shape[1]

    if frame_num == 1:
        # Ask user to mark parallel points and two points 6 feet apart. Order bl, br, tr, tl, p1, p2
        while True:
            image = frame
            cv2.imshow("image", image)
            cv2.waitKey(1)
            if len(mouse_pts) == 7:
                cv2.destroyWindow("image")
                break
            first_frame_display = False
        four_points = mouse_pts

        # Get perspective
        M, Minv = get_camera_perspective(frame, four_points[0:4])
        pts = src = np.float32(np.array([four_points[4:]]))
        warped_pt = cv2.perspectiveTransform(pts, M)[0]
        d_thresh = np.sqrt(
            (warped_pt[0][0] - warped_pt[1][0]) ** 2
            + (warped_pt[0][1] - warped_pt[1][1]) ** 2
        )
        bird_image = np.zeros(
            (int(frame_h * scale_h), int(frame_w * scale_w), 3), np.uint8
        )

        bird_image[:] = SOLID_BACK_COLOR
        pedestrian_detect = frame

    print("FRAME N: ", frame_num)

    # draw polygon of ROI
    pts = np.array(
        [four_points[0], four_points[1], four_points[3], four_points[2]], np.int32
    )
    cv2.polylines(frame, [pts], True, (0, 255, 255), thickness=4)

    # Detect person and bounding boxes using DNN
    pedestrian_boxes, num_pedestrians = DNN.detect_pedestrians(frame)

    if len(pedestrian_boxes) > 0:
        pedestrian_detect = plot_pedestrian_boxes_on_image(frame, pedestrian_boxes)
        warped_pts, bird_image = plot_points_on_bird_eye_view(
            frame, pedestrian_boxes, M, scale_w, scale_h
        )
        six_feet_violations, ten_feet_violations, pairs = plot_lines_between_nodes(
            warped_pts, bird_image, d_thresh
        )
        # plot_violation_rectangles(pedestrian_boxes, )
        total_pedestrians_detected += num_pedestrians
        total_pairs += pairs

        total_six_feet_violations += six_feet_violations / fps
        abs_six_feet_violations += six_feet_violations
        pedestrian_per_sec, sh_index = calculate_stay_at_home_index(
            total_pedestrians_detected, frame_num, fps
        )

    last_h = 75
    text = "ALARMAS < 1.5MTS: " + str(int(total_six_feet_violations))
    pedestrian_detect, last_h = put_text(pedestrian_detect, text, text_offset_y=last_h)

  #  text = "Stay-at-home Index: " + str(np.round(100 * sh_index, 1)) + "%"
  #  pedestrian_detect, last_h = put_text(pedestrian_detect, text, text_offset_y=last_h)

    if total_pairs != 0:
        sc_index = 1 - abs_six_feet_violations / total_pairs

  # text = "Social-distancing Index: " + str(np.round(100 * sc_index, 1)) + "%"
  #  pedestrian_detect, last_h = put_text(pedestrian_detect, text, text_offset_y=last_h)

    cv2.imshow("CCTVAL 2021", pedestrian_detect)
    cv2.waitKey(1)
    output_movie.write(pedestrian_detect)
    bird_movie.write(bird_image)


track.py
# limit the number of cpus used by high performance libraries
import os
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["OPENBLAS_NUM_THREADS"] = "1"
os.environ["MKL_NUM_THREADS"] = "1"
os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
os.environ["NUMEXPR_NUM_THREADS"] = "1"

import sys
sys.path.insert(0, './yolov5')

import argparse
import os
import platform
import shutil
import time
from pathlib import Path
import numpy as np
import cv2
import torch
import torch.backends.cudnn as cudnn

from yolov5.models.experimental import attempt_load
from yolov5.utils.downloads import attempt_download
from yolov5.models.common import DetectMultiBackend
from yolov5.utils.datasets import LoadImages, LoadStreams, VID_FORMATS
from yolov5.utils.general import (LOGGER, check_img_size, non_max_suppression, scale_coords,
                                  check_imshow, xyxy2xywh, increment_path, strip_optimizer, colorstr)
from yolov5.utils.torch_utils import select_device, time_sync
from yolov5.utils.plots import Annotator, colors, save_one_box
from deep_sort.utils.parser import get_config
from deep_sort.deep_sort import DeepSort

FILE = Path(__file__).resolve()
ROOT = FILE.parents[0]  # yolov5 deepsort root directory
if str(ROOT) not in sys.path:
    sys.path.append(str(ROOT))  # add ROOT to PATH
ROOT = Path(os.path.relpath(ROOT, Path.cwd()))  # relative
count = 0
count2 = 0
data = []

def detect(opt):
    out, source, yolo_model, deep_sort_model, show_vid, save_vid, save_txt, imgsz, evaluate, half, \
        project, exist_ok, update, save_crop = \
        opt.output, opt.source, opt.yolo_model, opt.deep_sort_model, opt.show_vid, opt.save_vid, \
        opt.save_txt, opt.imgsz, opt.evaluate, opt.half, opt.project, opt.exist_ok, opt.update, opt.save_crop
    webcam = source == '0' or source.startswith(
        'rtsp') or source.startswith('http') or source.endswith('.txt')

    # Initialize
    device = select_device(opt.device)
    half &= device.type != 'cpu'  # half precision only supported on CUDA

    # The MOT16 evaluation runs multiple inference streams in parallel, each one writing to
    # its own .txt file. Hence, in that case, the output folder is not restored
    if not evaluate:
        if os.path.exists(out):
            pass
            shutil.rmtree(out)  # delete output folder
        os.makedirs(out)  # make new output folder

    # Directories
    if type(yolo_model) is str:  # single yolo model
        exp_name = yolo_model.split(".")[0]
    elif type(yolo_model) is list and len(yolo_model) == 1:  # single models after --yolo_model
        exp_name = yolo_model[0].split(".")[0]
    else:  # multiple models after --yolo_model
        exp_name = "ensemble"
    exp_name = exp_name + "_" + deep_sort_model.split('/')[-1].split('.')[0]
    save_dir = increment_path(Path(project) / exp_name, exist_ok=exist_ok)  # increment run if project name exists
    (save_dir / 'tracks' if save_txt else save_dir).mkdir(parents=True, exist_ok=True)  # make dir

    # Load model
    model = DetectMultiBackend(yolo_model, device=device, dnn=opt.dnn)
    stride, names, pt = model.stride, model.names, model.pt
    imgsz = check_img_size(imgsz, s=stride)  # check image size

    # Half
    half &= pt and device.type != 'cpu'  # half precision only supported by PyTorch on CUDA
    if pt:
        model.model.half() if half else model.model.float()

    # Set Dataloader
    vid_path, vid_writer = None, None
    # Check if environment supports image displays
    if show_vid:
        show_vid = check_imshow()

    # Dataloader
    if webcam:
        show_vid = check_imshow()
        cudnn.benchmark = True  # set True to speed up constant image size inference
        dataset = LoadStreams(source, img_size=imgsz, stride=stride, auto=pt)
        nr_sources = len(dataset)
    else:
        dataset = LoadImages(source, img_size=imgsz, stride=stride, auto=pt)
        nr_sources = 1
    vid_path, vid_writer, txt_path = [None] * nr_sources, [None] * nr_sources, [None] * nr_sources

    # initialize deepsort
    cfg = get_config()
    cfg.merge_from_file(opt.config_deepsort)

    # Create as many trackers as there are video sources
    deepsort_list = []
    for i in range(nr_sources):
        deepsort_list.append(
            DeepSort(
                deep_sort_model,
                device,
                max_dist=cfg.DEEPSORT.MAX_DIST,
                max_iou_distance=cfg.DEEPSORT.MAX_IOU_DISTANCE,
                max_age=cfg.DEEPSORT.MAX_AGE, n_init=cfg.DEEPSORT.N_INIT, nn_budget=cfg.DEEPSORT.NN_BUDGET,
            )
        )
    outputs = [None] * nr_sources

    # Get names and colors
    names = model.module.names if hasattr(model, 'module') else model.names

    # Run tracking
    model.warmup(imgsz=(1 if pt else nr_sources, 3, *imgsz))  # warmup
    dt, seen = [0.0, 0.0, 0.0, 0.0], 0
    for frame_idx, (path, im, im0s, vid_cap, s) in enumerate(dataset):
        t1 = time_sync()
        im = torch.from_numpy(im).to(device)
        im = im.half() if half else im.float()  # uint8 to fp16/32
        im /= 255.0  # 0 - 255 to 0.0 - 1.0
        if len(im.shape) == 3:
            im = im[None]  # expand for batch dim
        t2 = time_sync()
        dt[0] += t2 - t1

        # Inference
        visualize = increment_path(save_dir / Path(path[0]).stem, mkdir=True) if opt.visualize else False
        pred = model(im, augment=opt.augment, visualize=visualize)
        t3 = time_sync()
        dt[1] += t3 - t2

        # Apply NMS
        pred = non_max_suppression(pred, opt.conf_thres, opt.iou_thres, opt.classes, opt.agnostic_nms, max_det=opt.max_det)
        dt[2] += time_sync() - t3

        # Process detections
        for i, det in enumerate(pred):  # detections per image
            seen += 1
            if webcam:  # nr_sources >= 1
                p, im0, _ = path[i], im0s[i].copy(), dataset.count
                p = Path(p)  # to Path
                s += f'{i}: '
                txt_file_name = p.name
                save_path = str(save_dir / p.name)  # im.jpg, vid.mp4, ...
            else:
                p, im0, _ = path, im0s.copy(), getattr(dataset, 'frame', 0)
                p = Path(p)  # to Path
                # video file
                if source.endswith(VID_FORMATS):
                    txt_file_name = p.stem
                    save_path = str(save_dir / p.name)  # im.jpg, vid.mp4, ...
                # folder with imgs
                else:
                    txt_file_name = p.parent.name  # get folder name containing current img
                    save_path = str(save_dir / p.parent.name)  # im.jpg, vid.mp4, ...

            txt_path = str(save_dir / 'tracks' / txt_file_name)  # im.txt
            s += '%gx%g ' % im.shape[2:]  # print string
            imc = im0.copy() if save_crop else im0  # for save_crop

            annotator = Annotator(im0, line_width=2, pil=not ascii)
            
            w, h = im0.shape[1], im0.shape[0]

            if det is not None and len(det):
                # Rescale boxes from img_size to im0 size
                det[:, :4] = scale_coords(im.shape[2:], det[:, :4], im0.shape).round()

                # Print results
                for c in det[:, -1].unique():
                    n = (det[:, -1] == c).sum()  # detections per class
                    if int(c) < len(names) :
                        s += f"{n} {names[int(c)]}{'s' * (n > 1)}, "  # add to string

                xywhs = xyxy2xywh(det[:, 0:4])
                confs = det[:, 4]
                clss = det[:, 5]

                # pass detections to deepsort
                t4 = time_sync()
                outputs[i] = deepsort_list[i].update(xywhs.cpu(), confs.cpu(), clss.cpu(), im0)
                t5 = time_sync()
                dt[3] += t5 - t4

                # draw boxes for visualization
                if len(outputs[i]) > 0:
                    for j, (output, conf) in enumerate(zip(outputs[i], confs)):

                        bboxes = output[0:4]
                        id = output[4]
                        cls = output[5]
                        count_obj(bboxes, w, h, id)

                        if save_txt:
                            # to MOT format
                            bbox_left = output[0]
                            bbox_top = output[1]
                            bbox_w = output[2] - output[0]
                            bbox_h = output[3] - output[1]
                            # Write MOT compliant results to file
                            with open(txt_path + '.txt', 'a') as f:
                                f.write(('%g ' * 10 + '\n') % (frame_idx + 1, id, bbox_left,  # MOT format
                                                               bbox_top, bbox_w, bbox_h, -1, -1, -1, i))

                            c = int(cls)  # integer class
                            if int(c) < len(names) :
                                label = f'{id} {names[c]} {conf:.2f}'                            
                                # Write results in an interpretable file
                                with open(txt_path + '_interpretable.csv', 'a') as f:
                                    f.write(f"{i},{frame_idx + 1},{label},{id},{bbox_left},{bbox_top},{bbox_w},{bbox_h}\n")
                                
                        if save_vid or save_crop or show_vid:  # Add bbox to image
                            c = int(cls)  # integer class
                            # label = f'{id} {names[c]} {conf:.2f}'
                            if int(c) < len(names) :
                                label = f'{names[c]}'
                                annotator.box_label(bboxes, label, color=colors(c, True))
                                if save_crop:
                                    txt_file_name = txt_file_name if (isinstance(path, list) and len(path) > 1) else ''
                                    save_one_box(bboxes, imc, file=save_dir / 'crops' / txt_file_name / names[c] / f'{id}' / f'{p.stem}.jpg', BGR=True)

                LOGGER.info(f'{s}Done. YOLO:({t3 - t2:.3f}s), DeepSort:({t5 - t4:.3f}s)')

            else:
                deepsort_list[i].increment_ages()
                LOGGER.info('No detections')

            # Stream results
            im0 = annotator.result()
            if show_vid:
                logo = cv2.imread('logo.png')
                size1 = 100
                size2 =150
                logo = cv2.resize(logo, (size2, size1))
                img2gray = cv2.cvtColor(logo, cv2.COLOR_BGR2GRAY)
                ret, mask = cv2.threshold(img2gray, 1, 255, cv2.THRESH_BINARY)
                roi = im0[-size1-10:-10, -size2-10:-10]
                roi[np.where(mask)] = 0
                roi += logo

                if opt.hcount:
                    color = (0, 255, 0)
                    start_point = (0, h - 350)
                    end_point = (int(w/2) - 50, h - 350)
                    cv2.line(im0, start_point, end_point, color, thickness=2)
                    org = (150, 150)
                    font = cv2.FONT_HERSHEY_COMPLEX
                    fontScale = 3
                    thickness = 3
                    cv2.putText(im0, str(count), org, font, fontScale, color, thickness, cv2.LINE_AA)

                    color = (255, 0, 0)
                    start_point = (int(w/2) + 50, h - 350)
                    end_point = (w, h - 350)
                    cv2.line(im0, start_point, end_point, color, thickness=2)
                    org = (w - 150, 150)
                    font = cv2.FONT_HERSHEY_COMPLEX
                    fontScale = 3
                    thickness = 3
                    cv2.putText(im0, str(count2), org, font, fontScale, color, thickness, cv2.LINE_AA)

                cv2.imshow(str(p), im0)
                cv2.waitKey(1)  # 1 millisecond

            # Save results (image with detections)
            if save_vid:
                if opt.hcount:

                    color = (0, 255, 0)
                    start_point = (0, h - 350)
                    end_point = (int(w/2) - 50, h - 350)
                    cv2.line(im0, start_point, end_point, color, thickness=2)
                    org = (150, 150)
                    font = cv2.FONT_HERSHEY_COMPLEX
                    fontScale = 3
                    thickness = 3
                    cv2.putText(im0, str(count), org, font, fontScale, color, thickness, cv2.LINE_AA)

                    color = (255, 0, 0)
                    start_point = (int(w/2) + 50, h - 350)
                    end_point = (w, h - 350)
                    cv2.line(im0, start_point, end_point, color, thickness=2)
                    org = (w - 150, 150)
                    font = cv2.FONT_HERSHEY_COMPLEX
                    fontScale = 3
                    thickness = 3
                    cv2.putText(im0, str(count2), org, font, fontScale, color, thickness, cv2.LINE_AA)

                if vid_path[i] != save_path:  # new video
                    vid_path[i] = save_path
                    if isinstance(vid_writer[i], cv2.VideoWriter):
                        vid_writer[i].release()  # release previous video writer
                    if vid_cap:  # video
                        fps = vid_cap.get(cv2.CAP_PROP_FPS)
                        w = int(vid_cap.get(cv2.CAP_PROP_FRAME_WIDTH))
                        h = int(vid_cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
                    else:  # stream
                        fps, w, h = 30, im0.shape[1], im0.shape[0]
                    save_path = str(Path(save_path).with_suffix('.mp4'))  # force *.mp4 suffix on results videos
                    vid_writer[i] = cv2.VideoWriter(save_path, cv2.VideoWriter_fourcc(*'mp4v'), fps, (w, h))
                vid_writer[i].write(im0)

    # Print results
    t = tuple(x / seen * 1E3 for x in dt)  # speeds per image
    LOGGER.info(f'Speed: %.1fms pre-process, %.1fms inference, %.1fms NMS, %.1fms deep sort update \
        per image at shape {(1, 3, *imgsz)}' % t)
    if save_txt or save_vid:
        s = f"\n{len(list(save_dir.glob('tracks/*.txt')))} tracks saved to {save_dir / 'tracks'}" if save_txt else ''
        LOGGER.info(f"Results saved to {colorstr('bold', save_dir)}{s}")
    if update:
        strip_optimizer(yolo_model)  # update model (to fix SourceChangeWarning)

def count_obj(box, w, h, id):
    global count, count2, data
    center_coor = (int(box[0] + (box[2]-box[0])/2 ), int(box[1] + (box[3] - box[1])/2 ))
    if int(box[1] + (box[3] - box[1])/2) > h - 350 and id not in data:
        if int(box[0] + (box[2]-box[0])/2) > int(w/2):
            count2 += 1
        else:
            count += 1
        data.append(id)
    

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('--yolo_model', nargs='+', type=str, default='yolov5m.pt', help='model.pt path(s)')
    parser.add_argument('--hcount', action='store_true', default=False, help='horizontal lines to count objects')  # file/folder, 0 for webcam
    parser.add_argument('--deep_sort_model', type=str, default='osnet_ibn_x1_0_MSMT17')
    parser.add_argument('--source', type=str, default='0', help='source')  # file/folder, 0 for webcam
    parser.add_argument('--output', type=str, default='inference/output', help='output folder')  # output folder
    parser.add_argument('--imgsz', '--img', '--img-size', nargs='+', type=int, default=[640], help='inference size h,w')
    parser.add_argument('--conf-thres', type=float, default=0.5, help='object confidence threshold')
    parser.add_argument('--iou-thres', type=float, default=0.5, help='IOU threshold for NMS')
    parser.add_argument('--fourcc', type=str, default='mp4v', help='output video codec (verify ffmpeg support)')
    parser.add_argument('--device', default='', help='cuda device, i.e. 0 or 0,1,2,3 or cpu')
    parser.add_argument('--show-vid', action='store_true', help='display tracking video results')
    parser.add_argument('--save-vid', action='store_true', help='save video tracking results')
    parser.add_argument('--save-txt', action='store_true', help='save MOT compliant results to *.txt')
    # class 0 is person, 1 is bycicle, 2 is car... 79 is oven
    parser.add_argument('--classes', nargs='+', type=int, help='filter by class: --class 0, or --class 16 17')
    parser.add_argument('--agnostic-nms', action='store_true', help='class-agnostic NMS')
    parser.add_argument('--augment', action='store_true', help='augmented inference')
    parser.add_argument('--update', action='store_true', help='update all models')
    parser.add_argument('--evaluate', action='store_true', help='augmented inference')
    parser.add_argument("--config_deepsort", type=str, default="deep_sort/configs/deep_sort.yaml")
    parser.add_argument("--half", action="store_true", help="use FP16 half-precision inference")
    parser.add_argument('--visualize', action='store_true', help='visualize features')
    parser.add_argument('--max-det', type=int, default=1000, help='maximum detection per image')
    parser.add_argument('--save-crop', action='store_true', help='save cropped prediction boxes')
    parser.add_argument('--dnn', action='store_true', help='use OpenCV DNN for ONNX inference')
    parser.add_argument('--project', default=ROOT / 'runs/track', help='save results to project/name')
    parser.add_argument('--name', default='exp', help='save results to project/name')
    parser.add_argument('--exist-ok', action='store_true', help='existing project/name ok, do not increment')
    opt = parser.parse_args()
    opt.imgsz *= 2 if len(opt.imgsz) == 1 else 1  # expand

    with torch.no_grad():
        detect(opt)
